library/alloc/src/boxed.rs

```diff
@@ -1325,10 +1325,11 @@ impl<T: Clone, A: Allocator + Clone> Clone for Box<T, A> {
 
 #[cfg(not(no_global_oom_handling))]
 #[stable(feature = "box_slice_clone", since = "1.3.0")]
-impl Clone for Box<str> {
+impl<A: Allocator + Clone> Clone for Box<str, A> {
     fn clone(&self) -> Self {
+        let alloc = Box::allocator(self).clone();
         // this makes a copy of the data
-        let buf: Box<[u8]> = self.as_bytes().into();
+        let buf: Box<[u8], A> = self.as_bytes().to_vec_in(alloc).into_boxed_slice();
```

Comment:
This is probably introducing extra checks that weren't here before since it uses into_boxed_slice which calls shrink_to_fit, whereas before we directly reused the buffer via the unsafe `into_box` I think. IMO we should keep that prior property, perhaps by allocating the Vec ourselves and copying bytes into it, then using into_box.

Reply:
Okay, I have rewritten it to use `RawVec::with_capacity_in` and `into_box`. I also implemented `clone_from` to re-use the allocation in the case that `self` and `other` are the same length, like `Clone for Box<[T], A>` does.